### PR TITLE
[FLINK-1417] Automatically register types with Kryo

### PR DESCRIPTION
**Pull request by rmetzger**
This pull request also contains code from old pull requests for FLINK-1391, FLINK-1392, FLINK-1395.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 1]
```diff
@@ -292,7 +360,76 @@ public void registerKryoType(Class<?> type) {
...
+		return disableAutoTypeRegistration;
+	}
+
+	/**
+	 * Control whether Flink is automatically registering all types in the user programs with
+	 * Kryo.
+	 *
+	 * @param disableAutoTypeRegistration
+	 */
+	public void setDisableAutoTypeRegistration(boolean disableAutoTypeRegistration) {
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java]: I would prefer "disableAutoTypeRegistration" here. And then "isAutoTypeRegistrationDisabled", above.

**Comment by aljoscha** [comment]
Looks good except for the one remark I had.

Did you run any performance tests on this? What is the overhead of using an Avro POJO vs. a plain POJO vs. using the Kryo Serializer?

**Comment by rmetzger** [comment]
Thank you for reviewing the pull request.

I've addressed your remark, added a fix for another issue (FLINK-1567) and now I'll run the performance tests.

**Comment by rmetzger** [comment]
The pull request now also contains changes to fix two classloading issues with the newly added `ExecutionConfig`.
Also, it fixes the Web interface for YARN.

I would like to merge this change soon. I've tested the code on extensively on a cluster and it seems to be stable.
I'm still working on the performance numbers.
The results I've got so far are the following:

Scale Factor 1 (lineitem only)
Pojo: 20
Kryo (with key selectors):  13
Avro (with key selectors): 10, 15, 13

Scale factor 100:
Pojo: 1634,

I'll update the comment to complete it.

**Comment by aljoscha** [comment]
What exactly are you running? TPC-H Query 3? Maybe we should test how fast Kryo would be with the PojoComparator.

**Comment by rmetzger** [comment]
No, its basically CoGrouping two "big" files (~50 and ~75 GB) 
https://github.com/project-flink/flink-perf/blob/master/flink-jobs/src/main/java/com/github/projectflink/avro/AvroPerf.java

**Comment by rmetzger** [comment]
I would like to merge this pull request soon.
@aljoscha, do you agree that we can investigate the performance for the PojoComparator also when the change is merged?

**Comment by aljoscha** [comment]
Yes please, go ahead. But if our Pojo stuff is really that slow we should think about how to improve that or remove it alltogether.

**Comment by rmetzger** [comment]
Yes.  First we need to understand why exactly the performance is so poor. Maybe its an issue we can "easily" fix.

**Comment by rmetzger** [comment]
Merging it now....

**Pull request closed without merging**